### PR TITLE
Module migration: Legacy type

### PR DESCRIPTION
**Pull request by ifarzana**
- Relates to https://github.com/xibosignageltd/xibo-private/issues/182

### CODE REVIEW

**Comment by ifarzana** [comment]
I am not sure about clock/countdown/wordclock compatibility classes. Do we need them?

File: lib/Factory/ModuleXmlTrait.php [hunk 1]
```diff
@@ -280,6 +281,27 @@ private function parseElements(\DOMNodeList $elementsNodes): array
...
+     * @return \Xibo\Widget\Definition\Property[]
+     */
+    private function parseLegacyTypes(\DOMNodeList $legacyTypeNodes): array
+    {
+        $legacyTypes = [];
+        foreach ($legacyTypeNodes as $node) {
+            /** @var \DOMNode $node */
+            if ($node instanceof \DOMElement) {
+                $legacyType = new LegacyType();
+                $legacyType->name = $node->textContent;
```
Comment (dasgarner) [inline lib/Factory/ModuleXmlTrait.php:295]: We should `trim()` this in case there are some stray characters in the nodes text content.

**Comment by dasgarner** [comment]
> I am not sure about clock/countdown/wordclock compatibility classes. Do we need them?

Ideally yes - we want to convert the widget->type on those from clock to the appropriate v4 module.

File: lib/Widget/ArticleWidgetCompatibility.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    public function upgradeWidget(Widget $widget, int $fromSchema, int $toSchema): void
+    {
+        $this->getLog()->debug('upgradeWidget: '. $widget->getId(). ' from: '. $fromSchema.' to: '.$toSchema);
+
+        foreach ($widget->widgetOptions as $option) {
+            $templateId = $widget->getOptionValue('templateId', '');
+
+            if ($option->option === 'templateId') {
+                switch ($templateId) {
+                    case 'media-rss-image-only':
```
[the remark this answers is not in the document]
Comment (ifarzana) [reply lib/Widget/ArticleWidgetCompatibility.php:46]: Shall I remove the others too from the xml file uch as media-rss-with-left-hand-text, media-rss-with-title etc?

[the remark this answers is not in the document]
Comment (dasgarner) [reply lib/Widget/ArticleWidgetCompatibility.php:46]: Yes absolutely - any that you have converted can be removed.

**Comment by ifarzana** [comment]
Can you please check WorldClock compatibity class, I am a bit unsure about it?

**Review by dasgarner — APPROVED** [review]
This looks great

**Pull request merged**